test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs

```diff
@@ -0,0 +1,3356 @@
+
+                    await resourceWriter.WriteStartAsync(tempCustomerResource);
+                    await resourceWriter.WriteEndAsync();
+                });
+
+                Assert.Equal(
+                    Strings.ValidationUtils_IncompatibleType("Collection(Edm.Int32)", "Collection(Edm.String)"),
+                    exception.Message);
+            }
+        }
+
+        [Fact]
+        public async Task WriterValidationUtils_NestedResourceInfoMustSpecifyIsCollectionAsync()
+        {
+            IODataRequestMessage requestMessage = new InMemoryMessage { Stream = this.stream };
+
+            await using (var messageWriter = new ODataMessageWriter(requestMessage, this.messageWriterSettings, this.model))
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(async () =>
+                {
+                    var tempCustomerNestedResourceInfo = new ODataNestedResourceInfo
+                    {
+                        Name = "Customer",
+                        IsCollection = false,
+                        IsComplex = false,
+                        Url = new Uri("http://tempuri.org/Orders(1)/Customer")
+                    };
+
+                    var resourceWriter = await messageWriter.CreateODataResourceWriterAsync(this.ordersEntitySet, this.orderEntityType);
+
+                    await resourceWriter.WriteStartAsync(this.orderResource);
+                    await resourceWriter.WriteStartAsync(tempCustomerNestedResourceInfo);
+                    // Set nested resource info IsCollection to null to cause subsequent WriteStartAsync(ODataResource) call to fail
+                    tempCustomerNestedResourceInfo.IsCollection = null;
+                    await resourceWriter.WriteStartAsync(this.customerResource);
+                    await resourceWriter.WriteEndAsync();
+                    await resourceWriter.WriteEndAsync();
+                    await resourceWriter.WriteEndAsync();
+                });
+
+                Assert.Equal(
+                    Strings.WriterValidationUtils_NestedResourceInfoMustSpecifyIsCollection("Customer"),
+                    exception.Message);
+            }
+        }
+
+        [Fact]
+        public async Task ODataMessageWriter_CannotWriteTopLevelNullAsync()
+        {
+            IODataResponseMessage responseMessage = new InMemoryMessage { Stream = this.stream };
+
+            await using (var messageWriter = new ODataMessageWriter(responseMessage, this.messageWriterSettings, this.model))
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(() => messageWriter.WritePropertyAsync(new ODataProperty
+                {
+                    Name = "Name",
+                    Value = null
+                }));
+
+                Assert.Equal(
+                    Strings.ODataMessageWriter_CannotWriteTopLevelNull,

```

Comment:
Why cannot write the Top level 'null'?

Reply:
@xuzhg Based on our logic (and the spec), if a top-level single-valued property is requested but it has a null value, the service should respond with 204 No Content
https://github.com/OData/odata.net/blob/713ea85bf61846a6978acc2a3d1c792e0a9d0af8/src/Microsoft.OData.Core/WriterValidator.cs#L253